bindings/matrix-sdk-ffi/src/room_directory_search.rs

```diff
@@ -0,0 +1,148 @@
+    pub is_world_readable: bool,
+    pub joined_members: u64,
+}
+
+impl From<matrix_sdk::room_directory_search::RoomDescription> for RoomDescription {
+    fn from(value: matrix_sdk::room_directory_search::RoomDescription) -> Self {
+        Self {
+            room_id: value.room_id.to_string(),
+            name: value.name,
+            topic: value.topic,
+            alias: value.alias.map(|alias| alias.to_string()),
+            avatar_url: value.avatar_url.map(|url| url.to_string()),
+            join_rule: PublicRoomJoinRule::convert(value.join_rule),
+            is_world_readable: value.is_world_readable,
+            joined_members: value.joined_members,
+        }
+    }
+}
+
+#[derive(uniffi::Object)]
+pub struct RoomDirectorySearch {
+    pub(crate) inner: RwLock<SdkRoomDirectorySearch>,
+}
+
+impl RoomDirectorySearch {
+    pub fn new(inner: SdkRoomDirectorySearch) -> Self {
+        Self { inner: RwLock::new(inner) }
+    }
+}
+
+#[uniffi::export(async_runtime = "tokio")]
+impl RoomDirectorySearch {
+    pub async fn next_page(&self) -> Result<(), ClientError> {
+        let mut inner = self.inner.write().await;
+        inner.next_page().await?;
+        Ok(())
+    }
+
+    pub async fn search(&self, filter: Option<String>, batch_size: u32) -> Result<(), ClientError> {
+        let mut inner = self.inner.write().await;
+        inner.search(filter, batch_size).await?;
+        Ok(())
+    }
+
+    pub async fn loaded_pages(&self) -> Result<u32, ClientError> {
+        let inner = self.inner.read().await;
+        Ok(inner.loaded_pages() as u32)
+    }
+
+    pub async fn is_at_last_page(&self) -> Result<bool, ClientError> {
+        let inner = self.inner.read().await;
+        Ok(inner.is_at_last_page())
+    }
+
+    pub async fn entries(
+        &self,
+        listener: Box<dyn RoomDirectorySearchEntriesListener>,
+    ) -> RoomDirectorySearchEntriesResult {
+        let entries_stream = self.inner.read().await.results();
+        RoomDirectorySearchEntriesResult {
+            entries_stream: Arc::new(TaskHandle::new(RUNTIME.spawn(async move {
+                // TODO: This needs to get improved with sensei Ivan
+                // pin_mut!(entries_stream);
+
+                // while let Some(diff) = entries_stream.next().await {
+                //     match diff {
+                //         VectorDiff::Clear => {
+                //             
+                // listener.on_update(RoomDirectorySearchEntryUpdate::Clear);
+                //         }
+                //         VectorDiff::Append { values } => {
+                //             
+                // listener.on_update(RoomDirectorySearchEntryUpdate::Append {
+                //                 values: values.into_iter().map(|value|
+                // value.into()).collect(),             });
+                //         }
+                //         _ => {}
+                //     }
+                // }
+            }))),
+        }
+    }
```

Comment:
Not tested, but it should work. You need the `RoomDirectorySearchEntriesResult` type anymore with this technique.

```suggestion
    pub async fn results(
        &self,
        listener: Box<dyn RoomDirectorySearchEntriesListener>,
    ) -> TaskHandle {
        let (initial_values, stream) = self.inner.read().await.results();
        
        TaskHandle::new(RUNTIME.spawn(async move {
            listener.on_update(vec![VectorDiff::Reset { values: initial_values }.map(Into::into)]);
            
            while let Some(diffs) = stream.next().await {
                listener.on_update(diffs.into_iter().map(|diff| diff.map(Into::into)).collect());
            }
        }))
    }
```